src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,168 @@
+            var n = 0;
+            do
+            {
+                var identifier = n > 0 ? n.ToString() : "";  // Make it look nice
+                codeBehindPath = Path.Combine(
+                    Path.GetDirectoryName(path),
+                    $"{Path.GetFileNameWithoutExtension(path)}{identifier}{Path.GetExtension(path)}.cs");
+                n++;
+            } while (File.Exists(codeBehindPath));
+
+            var codeBehindUri = new UriBuilder()
+            {
+                Scheme = Uri.UriSchemeFile,
+                Path = codeBehindPath,
+                Host = string.Empty,
+            }.Uri;
+
+            var className = Path.GetFileNameWithoutExtension(path);
+            var contents = text.GetSubTextString(new CodeAnalysis.Text.TextSpan(actionParams.ExtractStart, actionParams.ExtractEnd - actionParams.ExtractStart));
+            var compilationUnit = GenerateCodeBehindClass(className, contents, codeDocument);
+
+            var changes = new Dictionary<Uri, IEnumerable<TextEdit>>
+            {
+                [actionParams.Uri] = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = "",
+                        Range = codeDocument.RangeFromIndices(actionParams.RemoveStart, actionParams.RemoveEnd)
+                    }
+                },
+                [codeBehindUri] = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = compilationUnit.NormalizeWhitespace().ToFullString(),
+                        Range = new Range(new Position(0, 0), new Position(0, 0))
+                    }
+                }
+            };
+
+            var documentChanges = new List<WorkspaceEditDocumentChange>
+            {
+                new WorkspaceEditDocumentChange(new CreateFile() { Uri = codeBehindUri.ToString() })
+            };
+            
+            return new WorkspaceEdit()
+            {
+                Changes = changes,
+                DocumentChanges = documentChanges,
+            };
+        }
+
+        private IEnumerable<string> FindUsings(RazorCodeDocument razorCodeDocument)
+        {
+            return razorCodeDocument
+                .GetDocumentIntermediateNode()
+                .FindDescendantNodes<IntermediateNode>()
+                .Where(n => n is UsingDirectiveIntermediateNode)
+                .Select(n => ((UsingDirectiveIntermediateNode)n).Content);
+        }
+
+        private CompilationUnitSyntax GenerateCodeBehindClass(string className, string contents, RazorCodeDocument razorCodeDocument)
+        {
+            var namespaceNode = (NamespaceDeclarationIntermediateNode)razorCodeDocument
+                .GetDocumentIntermediateNode()
+                .FindDescendantNodes<IntermediateNode>()
+                .FirstOrDefault(n => n is NamespaceDeclarationIntermediateNode);
+
+            var @class = CSharpSyntaxFactory
```

Comment:
@jasonmalinowski @dibarbet @tmat is there a better way to be dynamically generating C# then what's done here? 

For context we're trying to generate a C# code behind file that corresponds to our Razor component document without access to the Roslyn workspace.

Reply:
/cc @CyrusNajmabadi as well